src/text_shaper/directwrite_shaper.cpp

```diff
@@ -140,70 +140,84 @@ optional<font_key> directwrite_shaper::load_font(font_description const& _descri
     std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> wStringConverter;
     std::wstring familyName = wStringConverter.from_bytes(_description.familyName);
 
-    for (UINT32 i = 0, e = fontCollection->GetFontFamilyCount(); i < e; ++i)
-    {
-        IDWriteFontFamily* family{};
-        fontCollection->GetFontFamily(i, &family);
+    UINT32 familyIndex;
+    BOOL familyExists = FALSE;
+    fontCollection->FindFamilyName(familyName.data(), &familyIndex, &familyExists);
+
+    if (!familyExists) {
+        // Fallback to Consolas
+        const wchar_t* consolas = L"Consolas";
+        fontCollection->FindFamilyName(consolas, &familyIndex, &familyExists);
+    }
 
-        IDWriteLocalizedStrings* names;
-        family->GetFamilyNames(&names);
+    Microsoft::WRL::ComPtr<IDWriteFontFamily> fontFamily;
+    fontCollection->GetFontFamily(familyIndex, &fontFamily);
 
-        BOOL exists = FALSE;
-        unsigned index{};
-        names->FindLocaleName(d->userLocale.c_str(), &index, &exists);
+    Microsoft::WRL::ComPtr<IDWriteLocalizedStrings> familyNames;
+    fontFamily->GetFamilyNames(&familyNames);
 
-        wchar_t name[64];
-        if (exists)
-            names->GetString(index, name, _countof(name));
-        else
-            // TODO: there was a different way of fallback that I have a tab open on
-            // one of my smartphones. Please find me.
-            names->GetString(i, name, sizeof(name));
+    BOOL localeExists = FALSE;
+    unsigned index{};
+    familyNames->FindLocaleName(d->userLocale.c_str(), &index, &localeExists);
 
-        if (familyName != name)
-            continue;
+    if (!localeExists) {
+        // Fallback to en-US
+        const wchar_t* localeName = L"en-US";
+        familyNames->FindLocaleName(localeName, &index, &localeExists);
+    }
 
-        for (UINT32 k = 0, ke = family->GetFontCount(); k < ke; ++k)
-        {
-            ComPtr<IDWriteFont> font;
-            family->GetFont(k, font.GetAddressOf());
+    if (!localeExists) {
+        // Fallback to locale where index = 0;
+        return nullopt;
+    }
+
+    UINT32 length = 0;
 
-            font_weight weight = dwFontWeight(font->GetWeight());
-            if (weight != _description.weight)
-                continue;
+    // TODO: very long font name?
+    wchar_t name[64];
+    familyNames->GetString(index, name, _countof(name));
```

Comment:
The `name` is currently unused now. But it *should* be saved into the `FontDescription` somewhere, because this name here might be different from the original name we ask DWrite to find. It's an edge case relatively, so I think it's OK to leave it here. Plus I don't know how to modify `FontDescription` because it seems read-only.

Reply:
Interesting. what's the difference of `_countof` to `sizeof`?